pkg/jx/cmd/step/syntax/step_syntax_canonical.go

```diff
@@ -0,0 +1,414 @@
+}
+
+func (o *StepSyntaxCanonicalOptions) createPipelineForKind(kind string, lifecycles *jenkinsfile.PipelineLifecycles, pipelines jenkinsfile.Pipelines, projectConfig *config.ProjectConfig, pipelineConfig *jenkinsfile.PipelineConfig) (*syntax.ParsedPipeline, error) {
+	var parsed *syntax.ParsedPipeline
+	var err error
+
+	if lifecycles != nil && lifecycles.Pipeline != nil {
+		parsed = lifecycles.Pipeline
+		for _, override := range pipelines.Overrides {
+			if override.MatchesPipeline(kind) {
+				parsed = syntax.ExtendParsedPipeline(parsed, override)
+			}
+		}
+	} else {
+		args := jenkinsfile.CreatePipelineArguments{
+			Lifecycles:        lifecycles,
+			PodTemplates:      o.PodTemplates,
+			CustomImage:       o.CustomImage,
+			DefaultImage:      o.DefaultImage,
+			WorkspaceDir:      o.getWorkspaceDir(),
+			GitHost:           o.GitInfo.Host,
+			GitName:           o.GitInfo.Name,
+			GitOrg:            o.GitInfo.Organisation,
+			ProjectID:         o.ProjectID,
+			DockerRegistry:    o.getDockerRegistry(projectConfig),
+			DockerRegistryOrg: o.GetDockerRegistryOrg(projectConfig, o.GitInfo),
+			KanikoImage:       o.KanikoImage,
+			NoKaniko:          o.NoKaniko,
+			// Make sure we don't inject the setversion steps
+			NoReleasePrepare: false,
+			StepCounter:      0,
+		}
+		parsed, _, err = pipelineConfig.CreatePipelineForBuildPack(args)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to generate pipeline from build pack")
+		}
+	}
+
+	parsed.AddContainerEnvVarsToPipeline(pipelineConfig.Env)
+
+	if pipelineConfig.ContainerOptions != nil {
+		mergedContainer, err := syntax.MergeContainers(pipelineConfig.ContainerOptions, parsed.Options.ContainerOptions)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Could not merge containerOptions from parent")
+		}
+		parsed.Options.ContainerOptions = mergedContainer
+	}
+
+	// TODO: Seeing weird behavior seemingly related to https://golang.org/doc/faq#nil_error
+	// if err is reused, maybe we need to switch return types (perhaps upstream in build-pipeline)?
+	ctx := context.Background()
+	if validateErr := parsed.Validate(ctx); validateErr != nil {
+		return nil, errors.Wrapf(validateErr, "validation failed for Pipeline")
+	}
+
+	return parsed, nil
+}
+
+func (o *StepSyntaxCanonicalOptions) combineEnvVars(projectConfig *jenkinsfile.PipelineConfig) error {
+	// add any custom env vars
+	envMap := make(map[string]corev1.EnvVar)
+	for _, e := range projectConfig.Env {
+		envMap[e.Name] = e
+	}
+	for _, customEnvVar := range o.CustomEnvs {
+		parts := strings.Split(customEnvVar, "=")
+		if len(parts) != 2 {
+			return errors.Errorf("expected 2 parts to env var but got %v", len(parts))
+		}
+		e := corev1.EnvVar{
+			Name:  parts[0],
+			Value: parts[1],
+		}
+		envMap[e.Name] = e
+	}
+	projectConfig.Env = syntax.EnvMapToSlice(envMap)
+	return nil
+}
+
+func (o *StepSyntaxCanonicalOptions) getWorkspaceDir() string {
+	return filepath.Join("/workspace", o.SourceName)
```

Comment:
Is /workspace really a given all the time?

Reply:
Yes, it is - that's always going to be the root directory in the Tekton step containers.

Reply:
I see. Maybe make "/workspace" a constant then? Something like _TEKTON_ROOT_DIR_?

Reply:
Sounds good. I'll do that shortly.